File: app/src/org/commcare/utils/DefaultCountryCodeSignalProvider.java
package org.commcare.utils;

import android.content.Context;
import android.telephony.TelephonyManager;

/**
 * Production implementation that reads country signals from Android system services.
 */
public class DefaultCountryCodeSignalProvider implements CountryCodeSignalProvider {
    private final TelephonyManager telephonyManager;
    private final Context context;

    public DefaultCountryCodeSignalProvider(Context context) {
        this.context = context;
        this.telephonyManager = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
    }

    @Override
    public String getSimCountryIso() {
        if (telephonyManager == null) {
            return "";
        }
        return telephonyManager.getSimCountryIso();
    }

    @Override
    public String getNetworkCountryIso() {
        if (telephonyManager == null) {
            return "";
        }
        return telephonyManager.getNetworkCountryIso();
    }

    @Override
    public String getLocaleCountry() {
        return context.getResources().getConfiguration().locale.getCountry();
    }
}

File: app/src/org/commcare/utils/DefaultCountryCodeSignalProvider.kt
package org.commcare.utils

import android.content.Context
import android.telephony.TelephonyManager
import androidx.core.os.ConfigurationCompat

/**
 * Production implementation that reads country signals from Android system services.
 */
class DefaultCountryCodeSignalProvider(context: Context) : CountryCodeSignalProvider {
    private val telephonyManager: TelephonyManager? =
        context.getSystemService(Context.TELEPHONY_SERVICE) as? TelephonyManager
    private val appContext: Context = context.applicationContext

    override val simCountryIso: String
        get() = telephonyManager?.simCountryIso ?: ""

    override val networkCountryIso: String
        get() = telephonyManager?.networkCountryIso ?: ""

    override val localeCountry: String
        get() = ConfigurationCompat.getLocales(appContext.resources.configuration)
            .get(0)?.country ?: ""
}
